MekHQ/src/mekhq/campaign/autoresolve/acar/handler/StandardUnitAttackHandler.java

```diff
@@ -0,0 +1,231 @@
+                    }
+                }
+            }
+        }
+    }
+
+    private int calculateDamage(Formation attacker, StandardUnitAttack attack,
+                                SBFUnit attackingUnit, Formation target) {
+        int damage = 0;
+        if (attack.getManeuverResult().equals(StandardUnitAttack.ManeuverResult.SUCCESS)) {
+            damage += 1;
+        }
+        damage += attackingUnit.getCurrentDamage().getDamage(attack.getRange()).damage;
+
+        damage = Math.max(1, processDamageByEngagementControl(attacker, target, damage));
+        return damage;
+    }
+
+    private void handleCrits(Formation target, SBFUnit targetUnit, int criticalRollResult, SBFUnit attackingUnit) {
+        if (criticalRollResult <= 4) {
+            // No crit
+            reporter.reportNoCrit();
+        } else if (criticalRollResult <= 7) {
+            // targeting critical
+            targetUnit.addTargetingCrit();
+            reporter.reportTargetingCrit(targetUnit);
+        } else if (criticalRollResult <= 9) {
+            // Damage crit
+            targetUnit.addDamageCrit();
+            reporter.reportDamageCrit(targetUnit);
+        } else if (criticalRollResult <= 11) {
+            // Both targeting and damage crit
+            targetUnit.addTargetingCrit();
+            targetUnit.addDamageCrit();
+            reporter.reportTargetingCrit(targetUnit);
+            reporter.reportDamageCrit(targetUnit);
+        } else {
+            // Destroyed
+            countKill(attackingUnit, targetUnit);
+            targetUnit.setCurrentArmor(0);
+            target.setHighStressEpisode();
+            reporter.reportUnitDestroyed();
+        }
+    }
+
+
+    private void countKill(SBFUnit attackingUnit, SBFUnit targetUnit) {
+
+        var killers = attackingUnit.getElements().stream().map(AlphaStrikeElement::getId)
+            .map(e -> simulationManager().getGame().getEntity(e)).filter(Optional::isPresent).map(Optional::get).toList();
+        var targets = targetUnit.getElements().stream().map(AlphaStrikeElement::getId)
+            .map(e -> simulationManager().getGame().getEntity(e)).filter(Optional::isPresent).map(Optional::get).toList();
+        for (var target : targets) {
+            RandomUtils.fastSample(killers).ifPresent(e -> e.addKill(target));
+        }
+    }
+
+    private int processDamageByEngagementControl(Formation attacker, Formation target, int damage) {
+        var engagementControlMemories = attacker.getMemory().getMemories("engagementControl");
+        var engagement = engagementControlMemories
+            .stream()
+            .filter(f -> f.getOrDefault("targetFormationId", Entity.NONE).equals(target.getId()))
+            .findFirst(); // there should be only one engagement control memory for this target
+
+        if (engagement.isEmpty()) {
+            return damage;
+        }
+
+        var isAttacker = (boolean) engagement.get().getOrDefault("attacker", false);
```

Comment:
I've noticed you use `var` a lot, is there value in narrowing down the type to specific uses? `var` becoming `boolean`, in this case?

I know `var` lets the compiler infer type, based on what comes after, but it feels like it's opening us up to type cast errors if we, for example, put a boolean in `isAttacker` and then later try and treat it as an incompatible class.

If this isn't a concern, I'd appreciate it if you could explain why. Purely for my own edification :)

Reply:
This isnt a concern, var is not like writting `?` or `Object` and then letting the compiler discover, its instead just a shorthand for the type of the variable, so here I could use `boolean isAttacker` or write down `var isAttacker` without problem because in both cases the IDE and the compiler knows I am using a boolean here.

The casting to (boolean) I am doing in this case is specifically because the `getOrDefault` is dealing with a variable that is a `Map<String, Object>`, so I need the typecasting here otherwise it will resolve the variable type to `Object`.

Reply:
Every day is a school day. Thanks for the explanation :)